### PR TITLE
Improve bitcode generation for Apple platforms

### PR DESCRIPTION
**Pull request by thombles**
Some improvements for iOS bitcode support suggested by Alex over at https://github.com/getditto/rust-bitcode/issues/9. r? @alexcrichton 

This improves Rust's bitcode generation so that provided you have a compatible LLVM version, Rust targeting iOS should work out of the box when compiled into bitcode-enabled apps, and when submitted to the App Store. I've tested these changes using Xcode 11.4.1 and Apple's vendored LLVM, [tag `swift-5.2.3-RELEASE`](https://github.com/apple/llvm-project/releases/tag/swift-5.2.3-RELEASE).

1. Force `aarch64-apple-ios` and `aarch64-apple-tvos` targets to always emit full bitcode sections, even when cargo is trying to optimise by invoking `rustc` with `-Cembed-bitcode=no`. Since Apple recommends bitcode on iOS and requires it on tvOS it is likely that this is what developers intend. Currently you need to override the codegen options with `RUSTFLAGS`, which is far from obvious.
2. Provide an LLVM cmdline in the target spec. Apple's bitcode verification process looks for some arguments. For Rust modules to be accepted we must pretend they were produced similarly. A suitable default is provided in `TargetOptions` for iOS, copied directly from the a clang equivalent section.

In the context of Apple platforms, the predominant purpose of bitcode is App Store submissions, so simulator and 32-bit targets are not relevant. I'm hoping that the cmdline strings will not be a maintenance burden to keep up-to-date. If the event of any future incompatibilities, hopefully a custom target config would offer enough flexibility to work around it. It's impossible to say for sure.

Due to unrelated build errors I haven't been able to build and test a full tvOS toolchain. I've stopped short of providing a similar `bitcode_llvm_cmdline` until I can actually test it.

### CODE REVIEW

File: src/librustc_codegen_ssa/back/write.rs [hunk 1]
```diff
@@ -147,6 +148,8 @@ impl ModuleConfig {
             || sess.opts.cg.linker_plugin_lto.enabled()
         {
             EmitObj::Bitcode
+        } else if sess.target.target.options.forces_embed_bitcode {
```
Comment (alexcrichton) [inline src/librustc_codegen_ssa/back/write.rs:151]: Could this get folded into `need_crate_bitcode_for_rlib`?

**Review by alexcrichton — APPROVED** [review]
Looks great to me, thanks! Just one minor thought but otherwise r=me

File: src/librustc_codegen_ssa/back/write.rs (hunk 1 above)
Comment (thombles) [reply src/librustc_codegen_ssa/back/write.rs]: This is interesting. Right now I can't see any code path for embedding `Marker` bitcode. Do we need it? IMO probably not. Would it make any sense for `OptLevel` to downgrade the wishes of my target spec from `Full` to `Marker`? Again probably not. It's certainly surprising, and it complicates the situation for someone debugging Rust in Xcode.

So at the risk of being over-enthusiastic, I've codified this by removing the `Marker` variant along with the ineffectual `OptLevel` match. Do you reckon this makes sense?

Comment (alexcrichton) [reply src/librustc_codegen_ssa/back/write.rs]: Oh thanks for the keen eye! I think that may have actually been an issue from an earlier one of my PRs, but at this point I think it's safe to remove. We can always add it back in later if truly necessary.

**Comment by vitiral** [comment]
Skimming PRs on TWiR, I read this as "Improve *bitcoin* generation for Apple platforms" :laughing:

**Pull request merged**